### ISSUE TITLE
Session.detach hangs

### ISSUE THREAD

**Issue by sc07kvm** [frida/frida#1056]
When calling session.detach from python code frida hangs:
```
(gdb) bt 
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
#1  0x00007f456b9ffc67 in g_cond_wait () at ../../../glib/glib/gthread-posix.c:1512
#2  0x00007f4568852d38 in frida_async_task_execute (self=self@entry=0x7f4544137800, cancellable=cancellable@entry=0x0, error=error@entry=0x7f44a5f98d68) at ../../../frida-core/src/frida.vala:2591
#3  0x00007f456885552a in frida_session_detach_sync (self=<optimized out>, cancellable=0x0, error=0x7f44a5f98db0) at ../../../frida-core/src/frida.vala:1799
#4  0x00007f4568824a55 in PySession_detach (self=0x7f45722d2b10) at ../../../frida-python/src/_frida.c:3061
#5  0x00007f4612280dd0 in _PyCFunction_FastCallDict (func_obj=func_obj@entry=0x7f44d107eca8, args=0x7f45d417d4e0, nargs=0, kwargs=<optimized out>) at /usr/src/debug/Python-3.6.8/Objects/methodobject.c:209
```
Pixel 2, Android 8.1.0, Frida 12.7.11

**Comment by sc07kvm** [comment]
`from typing import Set, List, Any, Optional, Dict
from threading import Thread, Event
import queue
import time

import frida
import frida.core

PATH_SCRIPT = "script.js"
PACKAGE = "com.konsierge.otkrytie.private"


class Hooking():
    def __init__(self, sn: str):
        self.sn = sn

        self.event_stop_frida_callback = Event()
        self.queue_frida_callback = queue.Queue()
        self.th_frida_callback = Thread(
            target=self._loop_frida_callback)
        self.th_frida_callback.daemon = True

        self.frida_device: frida.core.Device = None
        self.set_frida_sessions: Set[frida.core.Session] = set()
        self.set_frida_scripts: Set[frida.core.Script] = set()

    def _on_frida_callback(
            self,
            obj_name: str,
            func_name: str,
            params: List[Any]) -> None:
        self.queue_frida_callback.put((obj_name, func_name, params))

    def _loop_frida_callback(self) -> bool:
        while True:
            message = self.queue_frida_callback.get()
            if message is None and self.event_stop_frida_callback.is_set():
                break
            obj_name, func_name, params = message
            if obj_name == "device":
                if func_name == "spawn-added":
                    self._on_spawn_added(*params)
                elif func_name == "child-added":
                    self._on_child_added(*params)
            elif obj_name == "session":
                if func_name == "detached":
                    self._on_detached(*params)
            elif obj_name == "script":
                if func_name == "message":
                    self._on_message(*params)
                elif func_name == "destroyed":
                    self._on_destroyed(*params)
                elif func_name == "log":
                    self._on_log(*params)
        return True

    def start(self):
        self.th_frida_callback.start()
        self.frida_device = frida.get_device(self.sn)

        self.frida_device.on(
            "spawn-added", lambda spawn: self._on_frida_callback(
                "device", "spawn-added", [spawn]))
        self.frida_device.on(
            "child-added", lambda child: self._on_frida_callback(
                "device", "child-added", [child]))

        self.frida_device.enable_spawn_gating()

    def stop(self) -> bool:
        try:
            self.frida_device.disable_spawn_gating()
        except (
                frida.TransportError,
                frida.ServerNotRunningError,
                frida.InvalidOperationError):
            pass

        while True:
            try:
                frida_script = self.set_frida_scripts.pop()
            except KeyError:
                break
            try:
                frida_script.exports.end()
            except (
                    frida.TransportError,
                    frida.ServerNotRunningError,
                    frida.InvalidOperationError,
                    KeyError):
                pass

        while True:
            try:
                frida_session = self.set_frida_sessions.pop()
            except KeyError:
                break
            frida_session.detach()

        self.event_stop_frida_callback.set()
        self.queue_frida_callback.put(None)
        self.th_frida_callback.join()
        return True

    def inject(self, spawned_pid: int) -> bool:
        # session
        try:
            frida_session: frida.core.Session = \
                self.frida_device.attach(spawned_pid)
        except frida.ProcessNotFoundError:
            return False

        self.set_frida_sessions.add(frida_session)
        frida_session.on(
            "detached",
            lambda reason, crash: self._on_frida_callback(
                "session", "detached",
                [spawned_pid, frida_session, reason, crash]))

        # script
        with open(PATH_SCRIPT, "r") as f:
            script_source = f.read()

        try:
            frida_script = frida_session.create_script(script_source)
        except frida.InvalidArgumentError:
            return False
        self.set_frida_scripts.add(frida_script)
        frida_script.on(
            "message",
            lambda message, data: self._on_frida_callback(
                "script", "message", [
                    spawned_pid, frida_script,
                    message, data]))
        frida_script.on(
            "destroyed",
            lambda script: self._on_frida_callback(
                "script", "destroyed", [spawned_pid, frida_script]))
        frida_script.set_log_handler(
            lambda level, text: self._on_frida_callback(
                "script", "log", [spawned_pid, frida_script, level, text]))
        frida_script.load()
        try:
            info_script = frida_script.exports.init("config")
        except (frida.InvalidOperationError, KeyError):
            return False
        if info_script is None:
            return False
        return True

    def _on_spawn_added(
            self,
            spawn: frida._frida.Spawn  # pylint: disable=protected-access
                ) -> None:
        try:
            if spawn.identifier.startswith(PACKAGE):
                self.inject(spawn.pid)
            self.frida_device.resume(spawn.pid)
        except Exception:
            print("Error resume %r" % spawn)

    def _on_child_added(
            self,
            child: frida._frida.Child  # pylint: disable=protected-access
                ) -> None:
        try:
            self.frida_device.resume(child.pid)
        except Exception:
            print("Error resume %r" % child)

    # session
    def _on_detached(
            self, pid: int, session: frida.core.Session,
            reason, crash) -> None:
        self.set_frida_sessions.discard(session)

    # script
    def _on_message(
            self, _pid: int,
            _frida_script: frida.core.Script,
            _frida_prefix: int,
            message: Dict, data: Optional[bytes]):
        print("on_message", message, data)

    def _on_log(
            self, pid: int, frida_script: frida.core.Script,
            level, text):
        print("on_log", level, text)

    def _on_destroyed(self, pid: int, script: frida.core.Script):
        self.set_frida_scripts.discard(script)


def main():
    hook = Hooking("sn1234567")
    hook.start()
    time.sleep(600)
    hook.stop()


if __name__ == "__main__":
    main()
`
